include/property.h

```diff
@@ -0,0 +1,181 @@
+#ifndef __PROPERTY_H__
+#define __PROPERTY_H__
+
+#include <iostream>
+#include <set>
+#include <tuple>
+#include <sstream>
+#include <regex>
+
+namespace clayer {
+
+    // Concepts used for property reading and writing
+    template<typename T>
+    concept bool Streamable = requires(T o, std::ostream& s) {
+        {s << o} -> std::ostream&;
+    };
+
+    template<typename T>
+    concept bool StringRef = requires(T o) {
+        {*o} -> std::string;
+    };
+
+    // template<typename T=void>
+    // void print<>(){
+    //     std::cout << "\n";
+    // }
+
+    // template<Streamable T, typename... Targs>
+    // void print(T o, Targs... tail) {
+    //     std::cout << o <<" ";
+    //     print(tail...);
+    // }
+
+    // ## Log LogRecord - Any item that can be printed in a Log Record
+    enum log_properties {
+        FILE,   // - File name (CodeContext)
+        FUNC,   // - Function name (CodeContext)
+        LINE,   // - Line number (CodeContext)
+        // LEVEL
+        // NAME
+        // THREAD
+        // HASH
+        DATE,   // - Date a Log Record is created (RunContext)
+        TIME,   // - Time a Log Record is created (RunContext)
+        MESG    // - Added by the user
+    };
+
+    // ### Code Context - Log Properties that can be inferred from Code
+    struct CodeContext {
+        std::string file;
+        std::string func;
```

Comment:
(ContextInfo, not CodeInfo, right?) Sounds like a good plan to me - (no performance issue = use CodeContext in the logger).

If there is a big difference one technique I've seen is codebases keeping both a POD type (with plain char*s, no member functions) and a rich class like CodeContext, and a function for converting.

Google for example has a POD Time struct that just wraps an int64, for constexprs and fast organization, along with a fancy Time class that has all conversion functions and the formatted date texts stored.

If this is the case or we don't have time to go through the process you describe, I think since the analyser makes sense whether combined or separated from the actual logging class, 2 types is fine.